server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1706,10 +1752,33 @@ synchronized void addShardDeleteResult(
             ShardGeneration newGeneration,
             Collection<String> blobsToDelete
         ) {
+            if (compressed == null) {
+                // No output stream: skip writing, but still update generations
+                shardGenerationsBuilder.put(indexId, shardId, newGeneration);
+                return;
+            }
             try {
                 shardGenerationsBuilder.put(indexId, shardId, newGeneration);
-                new ShardSnapshotMetaDeleteResult(Objects.requireNonNull(indexId.getId()), shardId, blobsToDelete).writeTo(compressed);
-                resultCount += 1;
+
+                // Calculate how many bytes we want to write
+                int bytesToWriteIndexId = StreamOutput.bytesInString(indexId.getId());
+                int bytesToWriteShardId = StreamOutput.bytesInVInt(shardId);
+                int bytesToWriteBlobsToDelete = StreamOutput.bytesInStringCollection(blobsToDelete);
+                int totalBytesRequired = bytesToWriteIndexId + bytesToWriteShardId + bytesToWriteBlobsToDelete;
```

Comment:
I reckon even this much up-front computation is overly pedantic. The limit is in practice going to be 10s-of-MiBs (after compression) and the amount of data we write per entry is almost always going to be no more than a few kiB (before compression). There's also (a bounded amount of) buffering involved in the write process, both in the `BufferedOutputStream` and in zlib itself, which adds up to several kiB that's not counted in `shardDeleteResults.size()`.

Reply:
I can kinda see an argument for serializing `ShardSnapshotMetaDeleteResult#blobsToDelete` in a streaming fashion (i.e. not a length-prefixed array as it is today) so that we can check the limit before writing each blob name, truncating the collection if we hit the limit. I still think this is in YAGNI territory, these collections just aren't that large in practice anyway, but maybe if it's not too complex we can do that to put a firm constant bound on this memory usage?